cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreImpl.java

```diff
@@ -1421,77 +982,57 @@ public String extensionDataFolder(String extensionName) {
-                    if (newWinner.bodyModified || (newWinner.attachments != null && newWinner
-                            .attachments.hasChanged())) {
-
-                        // We need to work out which of the attachments for the revision are ones
-                        // we can copy over because they exist in the attachment store already and
-                        // which are new, that we need to prepare for insertion.
-                        Collection<Attachment> attachments = newWinner.getAttachments() != null ?
-                                newWinner.getAttachments().values() : new ArrayList<Attachment>();
-                        final List<PreparedAttachment> preparedNewAttachments =
-                                AttachmentManager.prepareAttachments(attachmentsDir,
-                                        attachmentStreamFactory,
-                                        AttachmentManager.findNewAttachments(attachments));
-                        final List<SavedAttachment> existingAttachments =
-                                AttachmentManager.findExistingAttachments(attachments);
-
-                        updateDocumentFromRevision(db, newWinner,
-                                preparedNewAttachments,
-                                existingAttachments);
-                    }
+            queue.submitTransaction(new ResolveConflictsForDocumentCallable(docTree, revIdKeep)).get();
 
 
-                    return null;
-                }
-            }).get();
+            // if this is a new or modified revision: graft the new revision on
+            if (newWinner.isBodyModified() || (newWinner.getAttachments() != null && ((ChangeNotifyingMap<String, Attachment>) newWinner
+                    .getAttachments()).hasChanged())) {
+
+                // We need to work out which of the attachments for the revision are ones
+                // we can copy over because they exist in the attachment store already and
+                // which are new, that we need to prepare for insertion.
+                Collection<Attachment> attachments = newWinner.getAttachments() != null ?
+                        newWinner.getAttachments().values() : new ArrayList<Attachment>();
+                final List<PreparedAttachment> preparedNewAttachments =
+                        AttachmentManager.prepareAttachments(attachmentsDir,
+                                attachmentStreamFactory,
+                                AttachmentManager.findNewAttachments(attachments));
+                final List<SavedAttachment> existingAttachments =
+                        AttachmentManager.findExistingAttachments(attachments);
+
+                queue.submit(new UpdateDocumentFromRevisionCallable(newWinner,
+                        preparedNewAttachments,
+                        existingAttachments, attachmentsDir, attachmentStreamFactory)).get();
+            }
+        } catch (AttachmentException e) {
+            // TODO
```

Comment:
Think we should at least add a log message here before merging even if a more robust solution is intended in a later pass.